### PR TITLE
Add support for locating mod sources with services from the classpath.

### PR DESCRIPTION
**Pull request by marchermans**
This allows a SERVICE-layer source set or jar to be loaded from the classpath in userdev if it is part of the MOD_CLASSES.

### CODE REVIEW

**Review by FiniteReality — APPROVED** [review]
This change seems fine as-is. Maybe we should consider extracting the parsing of MOD_CLASSES to a common location - I believe there's an exact duplicate of this logic in MinecraftLocator as well as one or two other places.

File: loader/src/main/java/net/neoforged/fml/loading/ClasspathTransformerDiscoverer.java [hunk 1]
```diff
@@ -59,4 +70,38 @@ private void locateTransformers(String resource) throws IOException {
...
+    private void scanModClasses() {
+        final Map<String, List<Path>> modClassPaths = getModClasses();
+        modClassPaths.forEach((modid, paths) -> {
+            if (shouldLoadInServiceLayer(paths)) {
+                found.add(new NamedPath(modid, paths.toArray(Path[]::new)));
+            }
+        });
+    }
+
+    private Map<String, List<Path>> getModClasses() {
```
Comment (Technici4n) [inline loader/src/main/java/net/neoforged/fml/loading/ClasspathTransformerDiscoverer.java:83]: There is no need to duplicate this, I think we can make the method in `CommonLaunchHandler` static and call it directly.

File: loader/src/main/java/net/neoforged/fml/loading/ClasspathTransformerDiscoverer.java [hunk 2]
```diff
@@ -59,4 +70,38 @@ private void locateTransformers(String resource) throws IOException {
             found.add(new NamedPath(path.toUri().toString(), path));
         }
     }
+
+    private void scanModClasses() {
+        final Map<String, List<Path>> modClassPaths = getModClasses();
+        modClassPaths.forEach((modid, paths) -> {
+            if (shouldLoadInServiceLayer(paths)) {
+                found.add(new NamedPath(modid, paths.toArray(Path[]::new)));
```
Comment (Technici4n) [inline loader/src/main/java/net/neoforged/fml/loading/ClasspathTransformerDiscoverer.java:78]: If a jar is loaded in the service layer like this, its modid should be blacklisted from being loaded by the normal mod locator. (similar to `ModDirTransformerDiscoverer#allExcluded()`)

**Review by Technici4n — CHANGES_REQUESTED** [review]

File: loader/src/main/java/net/neoforged/fml/loading/ClasspathTransformerDiscoverer.java (hunk 2 above)
Comment (marchermans) [reply loader/src/main/java/net/neoforged/fml/loading/ClasspathTransformerDiscoverer.java:78]: Right

File: loader/src/main/java/net/neoforged/fml/loading/TransformerDiscovererUtils.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import cpw.mods.jarhandling.JarContentsBuilder;
+import cpw.mods.jarhandling.JarMetadata;
+import cpw.mods.jarhandling.SecureJar;
+
+import java.nio.file.Path;
+
+public class TransformerDiscovererUtils {
+
+
+    public static boolean shouldLoadInServiceLayer(Path... path) {
```
Comment (Technici4n) [inline loader/src/main/java/net/neoforged/fml/loading/TransformerDiscovererUtils.java:17]: I'd put it in the same class as `TransformerDiscovererConstants`, potentially renaming it.

**Review by Technici4n — APPROVED** [review]

File: loader/src/main/java/net/neoforged/fml/loading/ModDirTransformerDiscoverer.java [hunk 4]
```diff
@@ -93,9 +90,6 @@ private static boolean shouldLoadInServiceLayer(Path path) {
         if (!path.toString().endsWith(".jar")) return false;
         if (LamdbaExceptionUtils.uncheck(() -> Files.size(path)) == 0) return false;
 
-        JarMetadata metadata = JarMetadata.from(new JarContentsBuilder().paths(path).build());
-        return metadata.providers().stream()
-            .map(SecureJar.Provider::serviceName)
-            .anyMatch(TransformerDiscovererConstants.SERVICES::contains);
+        return TransformerDiscovererConstants.shouldLoadInServiceLayer(path);
```
Comment (FiniteReality) [inline loader/src/main/java/net/neoforged/fml/loading/ModDirTransformerDiscoverer.java:93]: Now that we're delegating to this method, is `shouldLoadInServiceLayer` a good name for this method? Would something like `isValidServiceLayerJar` be a better name?

File: loader/src/main/java/net/neoforged/fml/loading/moddiscovery/MinecraftLocator.java [hunk 5]
```diff
@@ -47,7 +48,9 @@ public List<IModLocator.ModFileOrException> scanMods() {
                 .map(SecureJar::from)
                 .map(sj -> new ModFile(sj, this, ModFileParser::modsTomlParser))
                 .collect(Collectors.<IModFile>toList());
+        var otherModsExcluded = ClasspathTransformerDiscoverer.allExcluded();
         var othermods = baseMC.otherModPaths().stream()
+                .filter(p -> p.stream().noneMatch(otherModsExcluded::contains)) //We can no load MOD_CLASSES from the classpath if they are loaded on the SERVICE layer.
```
Comment (FiniteReality) [inline loader/src/main/java/net/neoforged/fml/loading/moddiscovery/MinecraftLocator.java:53]: Is `no load` a typo here?

File: loader/src/main/java/net/neoforged/fml/loading/ModDirTransformerDiscoverer.java (hunk 4 above)
Comment (Technici4n) [reply loader/src/main/java/net/neoforged/fml/loading/ModDirTransformerDiscoverer.java:93]: I think it's fine like this. Find a synonymous name doesn't make the method clearer.

**Review by Technici4n — APPROVED** [review]

**Pull request merged**